src/lib/clamp.jl

```diff
@@ -0,0 +1,64 @@
+
+import ZygoteRules: clamptype
+# This sees a tuple of argument types, and can modify the resulting tuple of tangents
+
+clamptype(Ts::Tuple{}, dxs::Tuple{}) = ()
+clamptype(Ts::Tuple, dxs::Tuple) =
+  first(Ts) === GlobalRef ? clamptype(Base.tail(Ts), dxs) :
+  (clamptype(first(Ts), first(dxs)), clamptype(Base.tail(Ts), Base.tail(dxs))...)
+
+clamptype(Ts::Tuple{}, dxs::Tuple) = (@error "mismatch!" dxs; dxs)
+clamptype(Ts::Tuple, dxs::Tuple{}) = (@error "mismatch!" Ts; ())
+
+# Bool, Real, Complex
+
+# clamptype(::Type{Bool}, dx) = nothing
+# clamptype(::Type{Bool}, dx::Complex) = nothing  # ambiguity
+# clamptype(::Type{<:AbstractArray{Bool}}, dx::AbstractArray) = nothing
+# clamptype(::Type{<:AbstractArray{Bool}}, dx::AbstractArray) = (@info "bool array" summary(dx); nothing)
+
+# clamptype(::Type{Bool}, dx) = (@info "bool, dropping" typeof(dx); nothing)
+# clamptype(::Type{Bool}, dx::Complex) = (@info "bool, dropping" typeof(dx); nothing)
+# clamptype(::Type{<:AbstractArray{Bool}}, dx::AbstractArray) = (@info "bool array, disabled" summary(dx); dx)
+
+clamptype(::Type{<:Real}, dx::Complex) = real(dx)
+clamptype(::Type{<:AbstractArray{<:Real}}, dx::AbstractArray) = real(dx)
+
+using LinearAlgebra: Diagonal, UpperTriangular, UnitUpperTriangular, LowerTriangular, UnitLowerTriangular
+using LinearAlgebra: AdjointAbsVec, TransposeAbsVec, AdjOrTransAbsVec
+
+# LinearAlgebra's matrix types
+
+for Wrap in [:Diagonal, :UpperTriangular, :LowerTriangular]
+  @eval begin
+    clamptype(::Type{<:$Wrap{T,PT}}, dx::$Wrap) where {T,PT} = 
+      clamptype(PT, dx)
+    clamptype(::Type{<:$Wrap{T,PT}}, dx::AbstractMatrix) where {T,PT} = 
+      clamptype(PT, $Wrap(dx))
+    # not right for :UnitUpperTriangular, :UnitLowerTriangular
+  end
+end
+
+for (trans, Wrap) in [(transpose, :Symmetric), (Base.adjoint, :Hermitian)]
+  @eval begin
+    clamptype(::Type{<:$Wrap{T,PT}}, dx::$Wrap) where {T,PT} = 
+      clamptype(PT, dx)
+    clamptype(::Type{<:$Wrap{T,PT}}, dx::AbstractMatrix) where {T,PT} = 
+      clamptype(PT, $Wrap(_twofold($trans, dx)))
+  end
+end
+
+_twofold(trans, dx) = (dx .+ trans(dx)) ./ 2
```

Comment:
Notice this:
```julia
using Zygote # this PR
using LinearAlgebra, FiniteDifferences, ForwardDiff, ChainRules

gradient(x -> Symmetric([x 3x; 5x 7x])[1,2], pi)   # (3,)
gradient(A -> A[1,2], Symmetric(rand(2,2)))        # ([0.0 0.5; 0.5 0.0],)
gradient(A -> Symmetric(A)[1,2], rand(2,2))        # ([0.0 1.0; 0.0 0.0],)

grad(central_fdm(5, 1), A -> A[1,2], Symmetric(rand(2,2))) # ([0.0 1.0; 1.0 0.0],) # weird? 
grad(central_fdm(5, 1), A -> Symmetric(A)[1,2], rand(2,2)) # ([0.0 1.0; 0.0 0.0],) # fine
```
If you use `_twofold(trans, dx) = dx .+ trans(dx) .- Diagonal(dx)`, then the Zygote results will double, and the 1st & 3rd will be quite obviously wrong. And this isn't a projection operator.

Is the FiniteDifferences result the right thing for a tangent not cotangent? Or is it a bug? As far as I can tell it's doing something like this:
```julia
ve, re = to_vec(Symmetric([1 3; 5 7])) # ve == [1,3,3,7]
re([1,-42,3,7]) == [1 3; 3 7]

re2(v) = Symmetric(reshape(v,2,2))
v2 = rand(1:999, 4); re(v2) == re2(v2)
re2(ForwardDiff.gradient(v -> re2(v)[1,2], ve)) == [0 1; 1 0]
```
Looking in ChainRules, there seem to be few tests using this behaviour of FiniteDifferences. There is a rule for `Matrix`, which applies `(dx .+ dx' .- Diagonal(dx)` to match:
```julia
rrule(Matrix, Symmetric(rand(2,2)))[2]([1 3; 5 7])[2]    == [1 8; 8 7] 
pullback(Matrix, Symmetric(rand(3,3)))[2]([1 3; 5 7])[1] == [1 4; 4 7] # with this PR
```
Defined here https://github.com/JuliaDiff/ChainRules.jl/blob/4e3164a3a48d4da35e0112d30be7ea9dbdaf3920/src/rulesets/LinearAlgebra/symmetric.jl#L71 where `_symmetric_back` is also the gradient of `Symmetric` where it makes more sense IMO. (Originally from Zygote, I think.)